backend/de.metas.marketing/activecampaign/src/main/java/de/metas/marketing/gateway/activecampaign/restapi/RestService.java

```diff
@@ -0,0 +1,166 @@
+
+import javax.annotation.Nullable;
+import java.net.URI;
+
+import static de.metas.marketing.gateway.activecampaign.ActiveCampaignConstants.ACTIVE_CAMPAIGN_API_TOKEN_HEADER;
+
+@Service
+public class RestService
+{
+	private static final Logger log = LogManager.getLogger(RestService.class);
+
+	private final ResponseErrorHandler responseErrorHandler;
+	private final RateLimitService rateLimitService;
+
+	public RestService(
+			@NonNull final ResponseErrorHandler responseErrorHandler,
+			@NonNull final RateLimitService rateLimitService)
+	{
+		this.responseErrorHandler = responseErrorHandler;
+		this.rateLimitService = rateLimitService;
+	}
+
+	public <T> ResponseEntity<T> performGet(@NonNull final ApiRequest getRequest, final Class<T> clazz)
+	{
+		final URI resourceURI;
+
+		final UriComponentsBuilder uriBuilder = UriComponentsBuilder.fromHttpUrl(getRequest.getBaseURL());
+
+		if (!Check.isEmpty(getRequest.getPathVariables()))
+		{
+			uriBuilder.pathSegment(getRequest.getPathVariables().toArray(new String[0]));
+		}
+
+		if (!Check.isEmpty(getRequest.getQueryParameters()))
+		{
+			uriBuilder.queryParams(getRequest.getQueryParameters());
+		}
+
+		final HttpEntity<String> request = new HttpEntity<>(getHttpHeaders(getRequest.getApiKey()));
+
+		resourceURI = uriBuilder.build().encode().toUri();
+
+		return performWithRetry(resourceURI, HttpMethod.GET, request, clazz);
+	}
+
+	public <T> ResponseEntity<T> performPost(@NonNull final ApiRequest postRequest, @Nullable final Class<T> clazz)
+	{
+		final URI resourceURI;
+
+		final UriComponentsBuilder uriBuilder = UriComponentsBuilder.fromHttpUrl(postRequest.getBaseURL());
+
+		if (!Check.isEmpty(postRequest.getPathVariables()))
+		{
+			uriBuilder.pathSegment(postRequest.getPathVariables().toArray(new String[0]));
+		}
+
+		if (!Check.isEmpty(postRequest.getQueryParameters()))
+		{
+			uriBuilder.queryParams(postRequest.getQueryParameters());
+		}
+
+		final String requestBody = postRequest.getRequestBody();
+
+		final HttpHeaders httpHeaders = getHttpHeaders(postRequest.getApiKey());
+
+		final HttpEntity<String> request = Check.isBlank(requestBody)
+				? new HttpEntity<>(httpHeaders)
+				: new HttpEntity<>(requestBody, httpHeaders);
+
+		resourceURI = uriBuilder.build().encode().toUri();
+
+		return performWithRetry(resourceURI, HttpMethod.POST, request, clazz);
+	}
+
+	private <T> ResponseEntity<T> performWithRetry(
+			final URI resourceURI,
+			final HttpMethod httpMethod,
+			final HttpEntity<String> request,
+			final Class<T> responseType)
+	{
+		ResponseEntity<T> response = null;
+		boolean retry = true;
+
+		while (retry)
+		{
+			try
+			{
+				log.debug("Performing [{}] on [{}] for request [{}]", httpMethod.name(), resourceURI, request);
+				retry = false;
+				response = createRestTemplate().exchange(resourceURI, httpMethod, request, responseType);
+			}
+			catch (final RateLimitExceededException rateLimitEx)
+			{
+				Loggables.withLogger(log, Level.ERROR)
```

Comment:
is this an error? if so, how do we make sure to not exceed the rate limit under normal circumstances?

Reply:
use `info` instead